File: cpp/sdk-basics/profiles/src/Profiles.cpp
/*
 * (c) 2018 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

#include "ISAgent.h"
#include "ISAgentSDKError.h"
#include <stdio.h>
#include <cstdlib>
#include <iostream>

#ifdef _WIN32
    #define HOMEVAR "USERPROFILE"
#else 
    #define HOMEVAR "HOME"
#endif

int main(int argc, char* argv[]) {

    int nErrorCode;

    // initialize agent with password persistor
    std::string persistorPath = "../../../sample-data/persistors/sample-persistor.pw";
    std::string persistorPassword = "ionic123";
    ISAgentDeviceProfilePersistorPassword persistor;
    persistor.setFilePath(persistorPath);
    persistor.setPassword(persistorPassword);
    ISAgent agent;
    nErrorCode = agent.initialize(persistor);
    if (nErrorCode != ISAGENT_OK) {
        std::cerr << "Failed to initialize agent from password persistor (" << persistorPath << ")" << std::endl;
        std::cerr << ISAgentSDKError::getErrorCodeString(nErrorCode) << std::endl;
        exit(1);
    }

    // verify that there are profiles
    if (!agent.hasAnyProfiles()) {
        std::cerr << "No profiles found in specified profile persistor" << std::endl;
        exit(1);
    }

    // list all available profiles
    std::cout << "ALL PROFILES:" << std::endl;
    const std::vector<ISAgentDeviceProfile> profiles = agent.getAllProfiles();
    for (int i=0; i < profiles.size(); i++) {
        const ISAgentDeviceProfile profile = profiles.at(i);
        std::cout << "---" << std::endl;
        std::cout << "ID       : " << profile.getDeviceId() << std::endl;
        std::cout << "Name     : " << profile.getName() << std::endl;
        std::cout << "Keyspace : " << profile.getKeySpace() << std::endl;
        std::cout << "ApiUrl   : " << profile.getServer() << std::endl;
    }

    // verify there is an active profile
    if (!agent.hasActiveProfile()) {
        std::cerr << "No profile set as active" << std::endl;
        exit(1);
    }
    
    // display active profile
    ISAgentDeviceProfile activeProfile = agent.getActiveProfile();
    std::cout << std::endl << "ACTIVE PROFILE: " << activeProfile.getDeviceId() << std::endl;

    // change active profile
    std::string newProfileId = "EfGh.1.54sdf8-sdfj-5802-sd80-248vwqucv9s73";
    std::cout << "\nSETTING NEW ACTIVE PROFILE: " << newProfileId << std::endl;
    bool bSuccess = agent.setActiveProfile(newProfileId);
    if (!bSuccess) {
        std::cerr << "Failed to set active profile" << std::endl;
        exit(2);
    }

    // display active profile
    ISAgentDeviceProfile newActiveProfile = agent.getActiveProfile();
    std::cout << std::endl << "NEW ACTIVE PROFILE: " << newActiveProfile.getDeviceId() << std::endl;

}

File: javascript/sdk-basics/profiles/profiles.js
/*
 * (c) 2018-2020 Ionic Security Inc.
 * By using this code, I agree to the Terms & Conditions (https://dev.ionic.com/use.html)
 * and the Privacy Policy (https://www.ionic.com/privacy-notice/).
 */

/*
 ** PRE-REQUISITES **
 * 1. A Machina tenant. You can obtain one at: https://ionic.com/start-for-free/.
 * 2. Your device, in this case the browser, needs to be enrolled. This is done 
 *    when you enrolled your device after signing up for a tenant. Enrollment 
 *    can also be accomplished by executing the Enroll Device script at:
 *    ../../enroll-device/index.html.
 */

import {getAgentConfig} from '../../jssdkConfig.js';

const main = async () => {
  
  // Get the tutorial application data. This assures all tutorils use the same
  // app ID, user ID and user authentication. It matches what was used for enrollment.
  const appData = getAgentConfig('Javascript Profiles Tutorial');
  let response;

  // Initialize the Machina agent.
  try {
    response = await new window.IonicSdk.ISAgent(appData);
  } catch (errorResp) {
    console.error('Error initializing ionic agent:' + errorResp.error);
  }
  const agent = response.agent;

  // Get all the profiles.
  try {
    response = await agent.queryProfiles(appData);
   } catch (errorResp) {
      console.log('Query profiles error: ' +  errorResp.error);
      return;
   }
  const profiles = response.profiles;

  // Verify there is at least one profile.
  if (profiles.length === 0) {
    console.log('No profiles found');
    return;
  }
    
  // List all available profiles.
  let index, len;
  console.log('');
  console.log('ALL PROFILES:');
  for (index = 0, len = profiles.length; index < len; ++index) {
    console.log(' ');
    console.log('Device ID : ' + profiles[index].deviceId);
    console.log('Created&nbsp;&nbsp; : ' + profiles[index].created);
    console.log('Keyspace&nbsp;  : ' + profiles[index].keyspace);
    console.log('ApiUrl&nbsp;&nbsp&nbsp; : ' + profiles[index].server);
  }
  console.log(' ');

  // Get the active profile.
  let activeProfile = true;
  try {
    response = await agent.getActiveProfile();
  } catch (errorResp) {
    console.log('Get Active Profile error: ' +  errorResp.error);
    activeProfile = false;
  }
  const activeDeviceId = response.deviceId;

  // Display active profile.
  if (activeProfile) {
    console.log('\nACTIVE PROFILE: ' + activeDeviceId);
  } else {
    console.log('\nNO ACTIVE PROFILE');
  }
  console.log(' ');
  
  // If there is an active profile and only one profile, then there is nothing to set.
  if (activeProfile && profiles.length === 1) {
    console.log('Only one profile, nothing to change.');
    return;
  }

  // Search or a non-active profile.
  let newProfileId = '';
  for (index = 0, len = profiles.length; index < len; ++index) {
    if (! profiles[index].active) {
      newProfileId = profiles[index].deviceId;
      break;
    }
  }

  if (newProfileId === '') {
    console.log('Didn\'t find a non-active profile');
    return;
  }

  // Change the active profile.
  console.log('\nSETTING NEW ACTIVE PROFILE: ' +  newProfileId);

  // Define the profile to make active.
  const profileToSet =
    { appId: appData.appId,
      userId: appData.userId,
      userAuth: appData.userAuth,
      deviceId: newProfileId
    };

  // Set the active profile.
  try {
    response = await agent.setActiveProfile(profileToSet);
  } catch (errorResp) { 
    console.log('Set active profile error: ' +  errorResp.error);
    return;
  }

  // Get the updated active profile.
  try {
    response = await agent.getActiveProfile();
  } catch (errorResp) {
    console.log('Get Active Profile error: ' +  errorResp.error);
    return;
  }
  const newActiveDeviceId = response.deviceId;

  // Display active profile.
  console.log('\nNEW ACTIVE PROFILE: ' + newActiveDeviceId);
};

main();
